### DISCUSSION TITLE
New generic manager "jsonpath", allowing json, yaml, toml, ..?

### DISCUSSION THREAD

**Discussion by LittleFox94** [renovatebot/renovate#14684] [Suggest an Idea]
Hi,

analog to the regex manager, but quite a bit easier to configure, a new manager reading common config file formats and extracting package name and version via jsonpaths could be useful

<details>
<summary>Example usage</summary>

Considering the files below, this manager would be able to find the git repository and docker images referenced in `my_config.yaml`, while being configurable for lots of different config files as long as they are in a supported format. In the example, the file format is being auto detected (by file extension or other heuristics) but a config to define it would probably be good.

```yaml
# my_config.yaml
dependencies:
- docker: nginx
  tag:    1.20.2
devDependencies:
- git:    https://github.com/renovatebot/renovate.git
  tag:    32.6.2
- docker: renovate/renovate
  tag:    32.6-slim
```

```jsonc
// renovate.json
{
  "configFileManagers": [
    {
      "fileMatch": ["^my_config.yaml$"],
      "entries": [
        ".dependencies",
        ".devDependencies"
      ],
      "match": [".git"], // could also be something like ".foo = 'bar'"
      "depName": ".git",
      "depVersion": ".tag",
      "datasourceTemplate": "git-tags",
    },
    {
      "fileMatch": ["^my_config.yaml$"],
      "entries": [
        ".dependencies",
        ".devDependencies"
      ],
      "match": [".docker"],
      "depName": ".docker",
      "depVersion": ".tag",
      "datasourceTemplate": "docker",
    }
  ]
}
```

The idea is to configure jsonpaths to a collection of "entries". If the found "entries" is a list, details for each element in it are extracted, if "entries" is an object, each key is considered an entry and `match`, `depName` and `depVersion` will operate on an object `{"key": $originalKey, "value": $originalValue }`.

`match` defines a check an entry has to pass before it's considered by this manger, in this case they check if the key `git` or `docker` is defined, using that as differentiation which `datasource` to use. Matches can be more complex, something like `".type = 'docker'"` could make sense if the config file to parse has a usable `type` field already. Of the list `match` only one element has to match for the entry to be considered; it's a logical `or`.

`depName` and `depVersion` are jsonpaths to extract dependency name and version from an entry.

Just to have another example, this manager could parse a `npm` `package.json` with a config like this:
```jsonc
// renovate.json
{
  "configFileManagers": [
    {
      "fileMatch": ["^package.json$"],
      "entries": [
        ".dependencies",
        ".devDependencies"
      ],
      "depName": ".key",
      "depVersion": .value",
      "datasourceTemplate": "npm",
    }
  ]
}
```
</details>

I'd try to implement that myself, if discussion is positive :)

**Comment by rarkins** [comment]
This would be really great! I'd love to see it.

BTW I think you could model it like "autodetect the file type based on extension, but allow manually specifying it"? that way if there's a file like `.renovaterc` which is actually JSON, you can still have a configuration like `"fileFormat": "json"`

**Reply by LittleFox94** [reply]
that's roughly what I meant with "but a config to define it would probably be good." :)

I'll look at implementing it, but can't promise anything - especially as I only use renovate at work for and not yet for any personal projects ^^'

**Reply by rarkins** [reply]
BTW it would be good to reuse the semantics from the regex manager as much as possible, because it seems to be working quite well so far.

**Reply by LittleFox94** [reply]
yep, we use that quite often - but I think all our uses could be replaced with this new one (lots of yaml and comment above the version field for the other values) ^^
